### PR TITLE
[Kernel] Update Kernel USER_GUIDE based on the latest API updates

### PR DESCRIPTION
**Pull request by vkorukanti**
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [x] Kernel
- [ ] Other (fill in here)

## Description
There have been a few changes to APIs since the last update to `USER_GUIDE.md`. Update it to reflect the latest API changes.

### CODE REVIEW

File: kernel/USER_GUIDE.md [hunk 1]
```diff
@@ -291,34 +292,40 @@ This method associates the information of a to-be-read file to a connector-speci
...
+* Data columns: Columns that are expected to be read from the Parquet file. Based on the `StructField` object defining the column, read the column in the Parquet file that matches the same name.
+* Metadata columns: These are special columns that must be populated using metadata about the Parquet file ([`StructField#isMetadataColumn`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/types/StructField.html#isMetadataColumn--) tells whether a column in `StructType` is a metadata column). To understand how to populate such a column, first match the column name against the set of standard metadata column name constants. For example, 
+    * `StructFileld#isMetadataColumn()` returns true and the column name is `StructField.ROW_INDEX_COLUMN_NAME`, then you have to a generate column vector populated with the actual index of each row in the Parquet file (that is, not indexed by the possible subset of rows return after Parquet data skipping).
```
Comment (allisonport-db) [inline kernel/USER_GUIDE.md:303]: Forgot TD wanted to update this variable name to include "metadata". I'll create a PR (I can update this in that PR too)

File: kernel/USER_GUIDE.md [hunk 2]
```diff
@@ -291,34 +292,40 @@ This method associates the information of a to-be-read file to a connector-speci
...
+The returned object is of type [`PredicateEvaluator`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/expressions/PredicateEvaluator.html). This is a special interface for evaluating Predicate on input batch returns a selection vector containing one value for each row in input batch indicating whether the row has passed the predicate or not. Optionally it takes an existing selection vector along with the input batch for evaluation. The result selection vector is combined with the given existing selection vector and a new selection vector is returned. This mechanism allows running an input batch through several predicate evaluations without rewriting the input batch to remove rows that do not pass the predicate after each predicate evaluation. The new selection should be the same or more selective as the existing selection vector. For example, if a row is marked as unselected in the existing selection vector, then it should remain unselected in the returned selection vector even when the given predicate returns true for the row.
```
Comment (allisonport-db) [inline kernel/USER_GUIDE.md:325]:
```suggestion
The returned object is of type [`PredicateEvaluator`](https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/expressions/PredicateEvaluator.html). This is a special interface for evaluating predicates on input batches. It returns a selection vector containing one value for each row in the input batch that indicates whether the row has passed the predicate or not. Optionally, it takes an existing selection vector along with the input batch for evaluation. The selection vector resulting from the predicate evaluation is combined with the given existing selection vector and a new selection vector is returned. This mechanism allows running an input batch through several predicate evaluations without rewriting the input batch to remove rows that do not pass the predicate after each predicate evaluation. The new selection should be the same or more selective as the existing selection vector. For example, if a row is marked as unselected in the existing selection vector, then it should remain unselected in the returned selection vector even when the given predicate returns true for the row.
```

**Review by allisonport-db — APPROVED** [review]
LGTM. We should make sure to double check the links in here after we are able to update the java docs.

**Comment by vkorukanti** [comment]
> LGTM. We should make sure to double check the links in here after we are able to update the java docs.

I used [this](https://github.com/tcort/markdown-link-check) tool to check whether the links are valid or not. I think this can be part of CI once we have a way to generate the docs as part of the CI builds.

```
(base) ➜  examples git:(userGuideUpdate) ✗ markdown-link-check ../USER_GUIDE.md

FILE: ../USER_GUIDE.md
  [✓] https://github.com/delta-io/delta/blob/master/PROTOCOL.md
  [✓] https://github.com/apache/spark
  [✓] https://github.com/apache/flink
  [✓] https://github.com/trinodb/trino
  [✓] #read-a-delta-table-in-a-single-process
  [✓] #build-a-delta-connector-for-a-distributed-processing-engine
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/client/TableClient.html
  [✓] https://github.com/delta-io/delta/tree/master/kernel/examples
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/Table.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/Snapshot.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/Scan.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/ScanBuilder.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/Scan.html#readData-io.delta.kernel.client.TableClient-io.delta.kernel.data.Row-io.delta.kernel.utils.CloseableIterator-java.util.Optional-
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/index.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-defaults/java/io/delta/kernel/defaults/client/DefaultTableClient.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/client/FileSystemClient.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/client/ParquetHandler.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/client/FileHandler.html#contextualizeFileReads-io.delta.kernel.utils.CloseableIterator-io.delta.kernel.expressions.Expression-
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/client/ParquetHandler.html#readParquetFiles-io.delta.kernel.utils.CloseableIterator-io.delta.kernel.types.StructType-
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/data/ColumnarBatch.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/data/ColumnVector.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/types/StructType.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/types/StructField.html#isMetadataColumn--
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/client/ExpressionHandler.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/client/ExpressionHandler.html#getEvaluator-io.delta.kernel.types.StructType-io.delta.kernel.expressions.Expression-io.delta.kernel.types.DataType-
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/expressions/ExpressionEvaluator.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/client/ExpressionHandler.html#createSelectionVector-boolean:A-int-int-
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/expressions/PredicateEvaluator.html
  [✓] https://docs.oracle.com/javaee/7/api/javax/resource/NotSupportedException.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/client/JsonHandler.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/client/JsonHandler.html#readJsonFiles-io.delta.kernel.utils.CloseableIterator-io.delta.kernel.types.StructType-
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/client/JsonHandler.html#parseJson-io.delta.kernel.data.ColumnVector-io.delta.kernel.types.StructType-
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/data/Row.html
  [✓] https://delta-io.github.io/delta/snapshot/kernel-api/java/io/delta/kernel/data/FilteredColumnarBatch.html

  34 links checked.
```

**Pull request merged**